apollo-at-link/src/join_spec_definition.rs

```diff
@@ -0,0 +1,319 @@
+use crate::argument::{
+    directive_optional_boolean_argument, directive_optional_enum_argument,
+    directive_optional_fieldset_argument, directive_optional_string_argument,
+    directive_required_enum_argument, directive_required_string_argument,
+};
+use crate::federation_schema::{FederationSchemaRef, OptionLinksMetadata};
+use crate::spec::{Identity, Url, Version};
+use crate::spec_definition::{SpecDefinition, SpecDefinitions};
+use apollo_compiler::schema::{Directive, DirectiveDefinition, EnumType, ExtendedType};
+use apollo_compiler::{Node, NodeStr};
+use apollo_federation_error::error::FederationError;
+use lazy_static::lazy_static;
+
+pub const JOIN_GRAPH_ENUM_NAME_IN_SPEC: &str = "Graph";
+pub const JOIN_GRAPH_DIRECTIVE_NAME_IN_SPEC: &str = "graph";
+pub const JOIN_TYPE_DIRECTIVE_NAME_IN_SPEC: &str = "type";
+pub const JOIN_FIELD_DIRECTIVE_NAME_IN_SPEC: &str = "field";
+pub const JOIN_IMPLEMENTS_DIRECTIVE_NAME_IN_SPEC: &str = "implements";
+pub const JOIN_UNIONMEMBER_DIRECTIVE_NAME_IN_SPEC: &str = "unionMember";
+pub const JOIN_ENUMVALUE_DIRECTIVE_NAME_IN_SPEC: &str = "enumValue";
+
+pub const JOIN_NAME_ARGUMENT_NAME: &str = "name";
+pub const JOIN_URL_ARGUMENT_NAME: &str = "url";
+pub const JOIN_GRAPH_ARGUMENT_NAME: &str = "graph";
+pub const JOIN_KEY_ARGUMENT_NAME: &str = "key";
+pub const JOIN_EXTENSION_ARGUMENT_NAME: &str = "extension";
+pub const JOIN_RESOLVABLE_ARGUMENT_NAME: &str = "resolvable";
+pub const JOIN_ISINTERFACEOBJECT_ARGUMENT_NAME: &str = "isInterfaceObject";
+pub const JOIN_REQUIRES_ARGUMENT_NAME: &str = "requires";
+pub const JOIN_PROVIDES_ARGUMENT_NAME: &str = "provides";
+pub const JOIN_TYPE_ARGUMENT_NAME: &str = "type";
+pub const JOIN_EXTERNAL_ARGUMENT_NAME: &str = "external";
+pub const JOIN_OVERRIDE_ARGUMENT_NAME: &str = "override";
+pub const JOIN_USEROVERRIDDEN_ARGUMENT_NAME: &str = "usedOverridden";
+pub const JOIN_INTERFACE_ARGUMENT_NAME: &str = "interface";
+pub const JOIN_MEMBER_ARGUMENT_NAME: &str = "interface";
+
+pub struct GraphDirectiveArguments<'a> {
+    pub name: &'a NodeStr,
+    pub url: &'a NodeStr,
+}
+
+pub struct TypeDirectiveArguments<'a> {
+    pub graph: &'a NodeStr,
+    pub key: Option<&'a NodeStr>,
+    pub extension: bool,
+    pub resolvable: bool,
+    pub is_interface_object: bool,
+}
+
+pub struct FieldDirectiveArguments<'a> {
+    pub graph: Option<&'a NodeStr>,
+    pub requires: Option<&'a NodeStr>,
+    pub provides: Option<&'a NodeStr>,
+    pub type_: Option<&'a NodeStr>,
+    pub external: Option<bool>,
+    pub override_: Option<&'a NodeStr>,
+    pub user_overridden: Option<bool>,
+}
```

Comment:
What's the reason for a lifetime for all of these structs to have a lifetime? Is it cause the getters for `NodeStr` in the compiler return a reference? Adding a lifetime to a struct means you'll have to carry it over for all of its uses, which makes a lot of the impls in this PR to have a whole bunch of lifetimes. I am afraid we are going to get ourselves in _a lot_ of trouble with these as we continue on this work.